File: leetcode/solutions/0525.ContiguousArray/findMaxLength.go
package solution

func findMaxLength(nums []int) int {
	var sum int
	var max int
	sumMap := make(map[int]int)

	for i, n := range nums {
		if n == 0 {
			sum--
		} else {
			sum++
		}

		if sum == 0 {
			max = i + 1
			continue
		}

		if index, ok := sumMap[sum]; !ok {
			sumMap[sum] = i
		} else if i-index > max {
			max = i - index
		}
	}

	return max
}

func maxInt(num1, num2 int) int {
	if num1 > num2 {
		return num1
	}

	return num2
}

File: leetcode/solutions/0525.ContiguousArray/findMaxLength.ts
export default function findMaxLength(nums: number[]): number {
  let sum = 0;
  let max = 0;
  const sumMap = new Map<number, number>();

  for (const [i, n] of nums.entries()) {
    if (n === 0) {
      sum--;
    } else {
      sum++;
    }

    if (sum === 0) {
      max = i + 1;
      continue;
    }

    if (!sumMap.has(sum)) {
      sumMap.set(sum, i);
    } else if (i - (sumMap.get(sum) ?? 0) > max) {
      max = i - (sumMap.get(sum) ?? 0);
    }
  }

  return max;
}
